File: internal/handler/webhooks/hook_actions.go
package webhooks

import (
	"github.com/gin-gonic/gin"
	"gitlab.com/gitlab-org/gitlab-redux/internal/model"
	"gitlab.com/gitlab-org/gitlab-redux/internal/service"
	"gitlab.com/gitlab-org/gitlab-redux/internal/util"
)

// HookActions handles webhook management actions
type HookActions struct {
	createService *service.WebHooksCreateService
	destroyService *service.WebHooksDestroyService
	logger        *util.Logger
}

// NewHookActions creates a new instance of HookActions
func NewHookActions(createService *service.WebHooksCreateService, destroyService *service.WebHooksDestroyService, logger *util.Logger) *HookActions {
	return &HookActions{
		createService: createService,
		destroyService: destroyService,
		logger:        logger,
	}
}

// Index displays a list of webhooks
func (h *HookActions) Index(ctx *gin.Context, relation model.WebHookRelation) {
	hooks := relation.SelectPersisted()
	hook := relation.New()

	ctx.HTML(200, "webhooks/index.html", gin.H{
		"hooks": hooks,
		"hook":  hook,
	})
}

// Create creates a new webhook
func (h *HookActions) Create(ctx *gin.Context, relation model.WebHookRelation) {
	currentUser, _ := ctx.Get("current_user")
	user, _ := currentUser.(model.User)

	hookParams := h.getHookParams(ctx)
	result := h.createService.Execute(user, hookParams, relation)

	if result.Success {
		ctx.Set("flash_notice", "Webhook created")
	} else {
		hooks := relation.SelectPersisted()
		ctx.Set("flash_alert", result.Message)
		ctx.HTML(200, "webhooks/index.html", gin.H{
			"hooks": hooks,
			"hook":  relation.New(),
		})
		return
	}

	ctx.Redirect(302, "/webhooks")
}

// Update updates an existing webhook
func (h *HookActions) Update(ctx *gin.Context, hook model.WebHook) {
	hookParams := h.getHookParams(ctx)

	if hook.Update(hookParams) {
		ctx.Set("flash_notice", "Webhook updated")
		ctx.Redirect(302, "/webhooks/"+hook.ID()+"/edit")
	} else {
		ctx.HTML(200, "webhooks/edit.html", gin.H{
			"hook": hook,
		})
	}
}

// Destroy deletes a webhook
func (h *HookActions) Destroy(ctx *gin.Context, hook model.WebHook) {
	h.destroyHook(ctx, hook)
	ctx.Redirect(302, "/webhooks")
}

// Edit displays the edit form for a webhook
func (h *HookActions) Edit(ctx *gin.Context, hook model.WebHook) {
	if hook == nil {
		ctx.Redirect(302, "/webhooks")
		return
	}

	hookLogs := h.getHookLogs(ctx, hook)

	ctx.HTML(200, "webhooks/edit.html", gin.H{
		"hook":     hook,
		"hookLogs": hookLogs,
	})
}

// getHookParams processes and returns the webhook parameters from the request
func (h *HookActions) getHookParams(ctx *gin.Context) map[string]interface{} {
	hook := ctx.PostForm("hook")
	if hook == "" {
		return nil
	}

	// Get the base hook parameters
	params := make(map[string]interface{})

	// Add the basic hook parameter names
	hookParamNames := []string{
		"enable_ssl_verification", "name", "description", "token", "url",
		"push_events_branch_filter", "branch_filter_strategy", "custom_webhook_template",
	}

	for _, name := range hookParamNames {
		if value := ctx.PostForm("hook[" + name + "]"); value != "" {
			params[name] = value
		}
	}

	// Handle token masking
	if ctx.Request.Method == "PUT" && params["token"] == model.WebHookSecretMask {
		delete(params, "token")
	}

	// Handle URL variables
	if urlVariables := ctx.PostFormArray("hook[url_variables][][key]"); len(urlVariables) > 0 {
		urlValues := ctx.PostFormArray("hook[url_variables][][value]")
		urlVars := make(map[string]string)

		for i, key := range urlVariables {
			if i < len(urlValues) && urlValues[i] != "" {
				urlVars[key] = urlValues[i]
			}
		}

		// For updates, merge with existing variables
		if ctx.Request.Method == "PUT" {
			existingHook, _ := ctx.Get("hook")
			if webHook, ok := existingHook.(model.WebHook); ok {
				existingVars := webHook.GetURLVariables()
				for k, v := range existingVars {
					if _, exists := urlVars[k]; !exists {
						urlVars[k] = v
					}
				}
			}
		}

		params["url_variables"] = urlVars
	}

	// Handle custom headers
	if customHeaders := ctx.PostFormArray("hook[custom_headers][][key]"); len(customHeaders) > 0 {
		customValues := ctx.PostFormArray("hook[custom_headers][][value]")
		customHeadersMap := make(map[string]string)

		for i, key := range customHeaders {
			if i < len(customValues) {
				value := customValues[i]

				// Check if we need to use the existing value from the database
				if value == model.WebHookSecretMask {
					existingHook, _ := ctx.Get("hook")
					if webHook, ok := existingHook.(model.WebHook); ok {
						existingHeaders := webHook.GetCustomHeaders()
						if existingValue, exists := existingHeaders[key]; exists {
							value = existingValue
						}
					}
				}

				customHeadersMap[key] = value
			}
		}

		params["custom_headers"] = customHeadersMap
	}

	return params
}

// destroyHook destroys a webhook
func (h *HookActions) destroyHook(ctx *gin.Context, hook model.WebHook) {
	currentUser, _ := ctx.Get("current_user")
	user, _ := currentUser.(model.User)

	result := h.destroyService.Execute(user, hook)

	if result.Status == "success" {
		if result.Async {
			ctx.Set("flash_notice", "Webhook scheduled for deletion")
		} else {
			ctx.Set("flash_notice", "Webhook deleted")
		}
	} else {
		ctx.Set("flash_alert", result.Message)
	}
}

// getHookLogs retrieves the logs for a webhook
func (h *HookActions) getHookLogs(ctx *gin.Context, hook model.WebHook) []model.WebHookLog {
	page := ctx.DefaultQuery("page", "1")
	return hook.GetWebHookLogs().Recent().Page(page).WithoutCount()
}

File: src/controllers/concerns/web_hooks/hook_actions.rs
use actix_web::{web, HttpResponse, Responder};
use serde::Deserialize;
use std::collections::HashMap;

use crate::models::web_hook::WebHook;
use crate::models::web_hook_log::WebHookLog;
use crate::services::web_hooks::{CreateService, DestroyService};

const SECRET_MASK: &str = "********";

#[derive(Debug, Deserialize)]
pub struct HookParams {
    enable_ssl_verification: Option<bool>,
    name: Option<String>,
    description: Option<String>,
    token: Option<String>,
    url: Option<String>,
    push_events_branch_filter: Option<String>,
    branch_filter_strategy: Option<String>,
    custom_webhook_template: Option<String>,
    url_variables: Option<Vec<UrlVariable>>,
    custom_headers: Option<Vec<CustomHeader>>,
}

#[derive(Debug, Deserialize)]
pub struct UrlVariable {
    key: String,
    value: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct CustomHeader {
    key: String,
    value: Option<String>,
}

pub trait HookActions {
    fn hooks(&self) -> &Vec<WebHook>;
    fn set_hooks(&mut self, hooks: Vec<WebHook>);
    fn hook(&self) -> &WebHook;
    fn set_hook(&mut self, hook: WebHook);
    fn current_user(&self) -> &User;
    fn relation(&self) -> &dyn WebHookRelation;

    fn index(&self) -> impl Responder {
        let hooks = self.relation().select_persisted();
        // In a real implementation, we'd need to handle the view rendering
        HttpResponse::Ok().finish()
    }

    fn create(&mut self, params: web::Json<HookParams>) -> impl Responder {
        let service = CreateService::new(self.current_user());
        let result = service.execute(params.into_inner(), self.relation());

        if result.is_success() {
            // In a real implementation, we'd need to handle flash messages
            HttpResponse::Found()
                .header("Location", "/web_hooks")
                .finish()
        } else {
            self.set_hooks(self.relation().select_persisted());
            // In a real implementation, we'd need to handle flash messages
            HttpResponse::Found()
                .header("Location", "/web_hooks")
                .finish()
        }
    }

    fn update(&mut self, params: web::Json<HookParams>) -> impl Responder {
        if self.hook().update(params.into_inner()) {
            // In a real implementation, we'd need to handle flash messages
            HttpResponse::Found()
                .header("Location", "/web_hooks/edit")
                .finish()
        } else {
            // In a real implementation, we'd need to handle the view rendering
            HttpResponse::Ok().finish()
        }
    }

    fn destroy(&self) -> impl Responder {
        self.destroy_hook(self.hook());
        HttpResponse::Found()
            .header("Location", "/web_hooks")
            .finish()
    }

    fn edit(&self) -> impl Responder {
        if self.hook().is_some() {
            // In a real implementation, we'd need to handle the view rendering
            HttpResponse::Ok().finish()
        } else {
            HttpResponse::Found()
                .header("Location", "/web_hooks")
                .finish()
        }
    }

    fn hook_logs(&self, page: i32) -> Vec<WebHookLog> {
        self.hook()
            .web_hook_logs()
            .recent()
            .page(page)
            .without_count()
    }

    fn destroy_hook(&self, hook: &WebHook) {
        let service = DestroyService::new(self.current_user());
        let result = service.execute(hook);

        match result.status {
            DestroyStatus::Success { async: true } => {
                // In a real implementation, we'd need to handle flash messages
            }
            DestroyStatus::Success { async: false } => {
                // In a real implementation, we'd need to handle flash messages
            }
            DestroyStatus::Error { message } => {
                // In a real implementation, we'd need to handle flash messages
            }
        }
    }
}

#[derive(Debug)]
pub enum DestroyStatus {
    Success { async: bool },
    Error { message: String },
}

pub trait WebHookRelation {
    fn select_persisted(&self) -> Vec<WebHook>;
}
